handler/ratelimit/limiter.go

```diff
@@ -125,22 +117,24 @@ func (l *Limiter) checkCapacity() (mode int, t time.Duration) {
 	now := time.Now()
 
 	for _, rl := range l.limits {
-		if rl.periodStart.IsZero() {
-			rl.periodStart = now
+		if rl.getPeriodStart().IsZero() {
+			rl.setPeriodStart(now)
 		}
-		
+
 		switch rl.window {
 		case windowFixed:
 			// Update current period.
-			multiplicator := ((now.UnixNano() - rl.periodStart.UnixNano()) / int64(time.Nanosecond)) / rl.period.Nanoseconds()
+			currentPeriod := rl.getPeriodStart()
+			multiplicator := ((now.UnixNano() - currentPeriod.UnixNano()) / int64(time.Nanosecond)) / rl.period.Nanoseconds()
 			if multiplicator > 0 {
-				rl.periodStart = rl.periodStart.Add(time.Duration(rl.period.Nanoseconds() * multiplicator))
-				rl.count = 0
+				currentPeriod = currentPeriod.Add(time.Duration(rl.period.Nanoseconds() * multiplicator))
+				rl.setPeriodStart(currentPeriod)
+				rl.count.Store(0)
```

Comment:
shouldn't we move this into `setPeriodStart()` (or maybe rename that to `resetPeriod()`. This way the two atomic values are not in a shared lock and multiple requests could override the count leading to some "free" requests.

Or is this prohibited by the non-buffered channel that will ever only let one request drip through the limiter (before fanning-out to the backend)?

Reply:
if so (serialization through channel), why do we have to use locks at all? If we have multiple rate-limiters in one backend they wouldn't share the same count and periodStart.

Reply:
Yes, it's actually serialized through channels. The initial change was to move the lock for all limiters (range) to specific ones, but this doesn't make sense either. I'll remove the lock too.